### PR TITLE
Feature/stop duplicate submissions

### PR DESCRIPTION
**Pull request by mattch1**
Code to prevent double submissions -using a  new flag in the session - this is added when the user check the company details and confirms. The flag is then used to determine whether the processor api should be called when the user later submits the application. 

If the application is not already submitted then the application can be submitted and the processor called, whereas if it already is, then it cannot be called a second time, the processor code is wrapped in a promise ".then" to ensure that the flag is set before submitting.

### CODE REVIEW

File: src/controllers/check.your.answers.controller.ts [hunk 1]
```diff
@@ -44,7 +44,7 @@ const route = async (req: Request, res: Response, next: NextFunction): Promise<v
 };
 
 export const submit = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
-  res.redirect(EXTENSIONS_CONFIRMATION);
+   res.redirect(EXTENSIONS_CONFIRMATION);
```
Comment (jterrych) [inline src/controllers/check.your.answers.controller.ts]: extra whitespace not required

File: src/controllers/company.details.controller.ts [hunk 2]
```diff
@@ -60,6 +62,12 @@ export const confirmCompanyStartRequest = async (req: Request, res: Response, ne
         logger.info(`User ${userProfile.id} has an extension request ` +
           `for companyNumber ${companyNumber}`);
       }
+      if (req.chSession.data[keys.SUBMITTED] === undefined) {
+        req.chSession.appendData(keys.SUBMITTED, false);
+      } else {
+        req.chSession.data[keys.SUBMITTED] = false;
```
Comment (jterrych) [inline src/controllers/company.details.controller.ts]: wont this line alone suffice?

File: src/controllers/confirmation.controller.ts [hunk 3]
```diff
@@ -25,20 +26,28 @@ const route = async (req: Request, res: Response, next: NextFunction): Promise<v
...
-    }
-    if (activeFeature(process.env.FEATURE_REQUEST_COUNT)) {
-      RequestCountMonitor.updateTodaysRequestNumber(1);
-    } else {
-      logger.info("Feature flag is toggled off for request number counter update");
-    }
+  const isSubmitted: boolean = req.chSession.data[keys.SUBMITTED];
+  if (!isSubmitted) {
+    req.chSession.data[keys.SUBMITTED] = true;
+    await saveSession(req.chSession).then(async () => {
```
Comment (jterrych) [inline src/controllers/confirmation.controller.ts]: the session save should probably happen after the client is called. if theres an error with the processor then they wont be able to try again until the session is cleared

File: src/controllers/confirmation.controller.ts [hunk 4]
```diff
@@ -25,20 +26,28 @@ const route = async (req: Request, res: Response, next: NextFunction): Promise<v
...
-      return next(e);
-    }
-    if (activeFeature(process.env.FEATURE_REQUEST_COUNT)) {
-      RequestCountMonitor.updateTodaysRequestNumber(1);
-    } else {
-      logger.info("Feature flag is toggled off for request number counter update");
-    }
+  const isSubmitted: boolean = req.chSession.data[keys.SUBMITTED];
+  if (!isSubmitted) {
+    req.chSession.data[keys.SUBMITTED] = true;
```
Comment (jterrych) [inline src/controllers/confirmation.controller.ts]: `req.chSession.data[keys.SUBMITTED]` equates to `submitted`. This provides no context in the global session. We should be putting this within the extensions session or at the very least renaming it to something less ambiguous

File: src/controllers/confirmation.controller.ts (hunk 3 above)
Comment (mattch1) [reply src/controllers/confirmation.controller.ts]: It depends how likely this is; I'd like to ensure as much as possible that no second click could trigger another processor especially now that CHIPS call is in there.

**Comment by jterrych** [comment]
lets get some unit tests up in here. We need too verify that under the right conditions it is indeed only called once. we need to ensure that an error thrown by the processor will not set submitted to true

File: src/controllers/confirmation.controller.ts (hunk 3 above)
Comment (jterrych) [reply src/controllers/confirmation.controller.ts]: Thats trading a minor inconvenience with potentially a nasty bug and is definitely not worth he trade.

**Pull request merged**